DM-25785: Add XMLlint format check, reformat XML files with xmllint --format

pkubanek opened the pull request.

r-owen reviewed (commented):
What does the github action do?

I suggest adding a git pre-commit hook that runs on the developers machine and prevents committing unless the XML files are all linted. I suggest using Python pre-commit for that; it seems very capable and it's a lot saner than trying to write a pre-commit hook yourself (as I have found to may embarrassment and regret; the black pre-commit hook in ts_salobj still has problems.)

Please also explain in the README (or somewhere) how to run the XML linter on the command line.

rbovill reviewed (changes requested):
Also, this PR is now a bit old and out of date with develop.  Could you rebase to pick up the recent changes?

pkubanek commented:
Rebased, added pre-commit hook with tests if XMLlint is installed, added description

tribeiro commented:
Can this PR wait for the next xml release?

r-owen commented on .githooks/pre-commit:
```diff
@@ -1,9 +1,31 @@
-if xmllint --version 2>/dev/null; then
-  echo "Linting XML files"
-  for f in $(find . -name *.xml); do xmllint --format $f > /tmp/$$lint; cp /tmp/$$lint $f; done && rm /tmp/$$lint
+
```
Please change above so that your hook runs `black --check .` on the same files you are running your xml linter on, and rejecting the commit if the check fails (meaning that one or more files would be reformatted).

r-owen commented on README.md:
````diff
@@ -41,6 +41,22 @@ from anywhere in the repository.  This will run the tests and print out a short
+xmllint (http://xmlsoft.org) is used in GitHub action to check committed files format. The files can be reformatted running xmllint --format, such as:
+
+```bash
+for f in $(find . -name *.xml); do xmllint --format $f > /tmp/$$lint; cp /tmp/$$lint $f; done && rm /tmp/$$lint
````
Thanks. I wish it wasn't so complicated, but I'm glad to know what to do

pkubanek replied:
Should  be fixed now.

r-owen reviewed (changes requested):
My one request is to modify the pre-commit hook to also run `black --check .` on the same commit that you are running `xmllint` and reject the commit if either changes. Right now the pre-commit hook reformats the Python code, needs to change.

pkubanek commented:
That can definitely wait for next XML release. Just let me know when XML is released, so I will trigger this PR again.

rbovill commented:
@pkubanek Once the PR is cleaned up (conflicts resolved and Russell's comments) this PR can be merged to develop and will be included in the next release.  Tiago just meant it was too late to get it into today's release.

pkubanek commented:
Ready to commit. Fixed black, added hopefully helpful messages what to do on problems to githooks. Please let me merge it before others start to modify XML files, so I don't have to rebase *(I put xmllinting as last commit, so I can do it much quicker than before, just in case..)*

r-owen reviewed (approved):
Thank you for fixing black handling in the pre-commit hook.
